### PR TITLE
Smooth heightmap collisions

### PR DESCRIPTION
**Pull request by agrande**
This PR implements smooth heightmap collisions:
- Not getting stuck when walking towards a rock section
- Sliding when jumping on a rock section
- Invisible walls preventing you from falling off cliffs (as long as you don't jump)

It comes with some optimizations for previous stuff like ground height detection and magicball bouncing.
I also added a few utility functions for formatting vectors and actor state in a readable way (when you want to display them for every frame, for instance), which proved to be very helpful when debugging.

**Preview here:** https://pr-585.lba2remake.net

### CODE REVIEW

**Comment by xesf** [comment]
Had a quick look and its definitely better on most cases.

There are still places where Twinsen gets stuck even without a slope. It does not walk in any directions. This happens a lot if you go play in the back of Citadel near Rosa/Cliffs.
Going down seems way better than before. Going up sometimes gets stuck on places you see it show go up a bit more.

It still allows to step on some bricks of 1 or 2 height high. I believe for bricks we should check the shape of the brick and if it's not any of the slope ones to block going forward. For example, you can step on the first terminal you encounter on LBA1 prison, and the stairs seem to have the same issue just going up.

**Comment by xesf** [comment]
Also notice actors can only be pushed in one direction (X) instead of (Z).
This is what's happen on master.

**Comment by slwilliams** [comment]
one thing we should do is update the logic that saves the position to reset Twinsen when he drowns to not consider a location safe if he's sliding. Should be trivial to update validPosition in hero.ts.

One bug I found was getting stuck on objects: 
image

not sure it's worth fixing though, isn't going to happen often.

**Review by slwilliams — APPROVED** [review]

**Comment by agrande** [comment]
> one thing we should do is update the logic that saves the position to reset Twinsen when he drowns to not consider a location safe if he's sliding. Should be trivial to update validPosition in hero.ts.


I already did that, if you look at the diff, I added !isSliding and !isStuck to validPosition

There's still some issue though when sliding down into the water, because I suspect that there's one frame at some point where you're not sliding anymore but not drowning yet.
Not sure how easy it is to fix, I couldn't find a simple solution. But I'll address it later as I don't think this PR is making things worse than they were before.

> One bug I found was getting stuck on objects: 
> 
> image
> 
> 
> 
> not sure it's worth fixing though, isn't going to happen often.

I'll take a look at this. It doesn't happen in the original?

**Comment by slwilliams** [comment]
ok testing in the original the trick with the emerald moon issue is that they have a timeout on the sliding. So if Twinsen gets stuck sliding and hasn't moved in x seconds they cancel the slide and allow you to jump again. This means in the case above you'd get stuck and then the sliding would be cancelled and then you can jump again to unstick yourself.

**Comment by agrande** [comment]
> ok testing in the original the trick with the emerald moon issue is that they have a timeout on the sliding. So if Twinsen gets stuck sliding and hasn't moved in x seconds they cancel the slide and allow you to jump again. This means in the case above you'd get stuck and then the sliding would be cancelled and then you can jump again to unstick yourself.

I actually did the sliding timeout too... Suprirsed it doesn't work in that case.

**Review by xesf — APPROVED** [review]

File: src/game/Extra.ts [hunk 1]
```diff
@@ -105,6 +105,11 @@ export default class Extra {
     hasCollidedWithActor: boolean;
     throwBy: number = -1;
     targetActor: number = -1;
+    /**
+     * This is used to add data meant to be looked
+     * at in the inspector for debugging purposes.
+     */
+    dbg: any = {};
```
[the remark this answers is not in the document]
Comment (agrande) [reply src/game/Extra.ts:112]: I would keep it. I find that I frequently need to display some values related to Actor | Extra in the inspector, whatever feature I'm working on. It's convenient to have something like that (similar to Three.js' userData property) where I can easily add and remove stuff while debugging something or prototyping a new feature, without being bothered by the compiler complaining that the field does not exist on type Actor or Extra.

[the remark this answers is not in the document]
Comment (agrande) [reply src/game/Extra.ts:112]: But if it bothers you I can remove it.

[the remark this answers is not in the document]
Comment (xesf) [reply src/game/Extra.ts:112]: Nop, not at all, was just curious. We can leave it there.

[the remark this answers is not in the document]
Comment (agrande) [reply src/game/Extra.ts:112]: Ok then, I'll merge as it is.

**Pull request merged**